src/Storages/ObjectStorageQueue/ObjectStorageQueueMetadataFactory.cpp

```diff
@@ -14,19 +14,26 @@ ObjectStorageQueueMetadataFactory & ObjectStorageQueueMetadataFactory::instance(
     return ret;
 }
 
-ObjectStorageQueueMetadataFactory::FilesMetadataPtr
-ObjectStorageQueueMetadataFactory::getOrCreate(const std::string & zookeeper_path, const ObjectStorageQueueSettings & settings)
+ObjectStorageQueueMetadataFactory::FilesMetadataPtr ObjectStorageQueueMetadataFactory::getOrCreate(
+    const std::string & zookeeper_path,
+    ObjectStorageQueueMetadataPtr metadata,
+    ObjectStorageQueueSettings & settings)
 {
     std::lock_guard lock(mutex);
     auto it = metadata_by_path.find(zookeeper_path);
     if (it == metadata_by_path.end())
     {
-        auto files_metadata = std::make_shared<ObjectStorageQueueMetadata>(zookeeper_path, settings);
-        it = metadata_by_path.emplace(zookeeper_path, std::move(files_metadata)).first;
+        metadata->syncWithKeeper();
```

Comment:
This sync with keeper previously was after getOrCreate in ctor of StorageObjectStorageQueue, but I moved it here, because it is more atomic (previously if checkEquals with keeper metadata failed it would be removed from Factory in a separate method, which means that someone could have fetched it from factory in the mean time, leading to incorrect behaviour)